File: LEETCODE/trees/easy/S3_InorderIteration.java
package LEETCODE.trees.easy;

import java.util.ArrayList;
import java.util.*;

import LEETCODE.trees.BinaryTree;

public class InorderIteration {

    private static List<Integer> inorder(BinaryTree.Node root){
        List<Integer> ans  = new ArrayList<>();
        Stack<BinaryTree.Node> stack = new Stack<>();
        
        if(root == null){
            return ans;
        }

        BinaryTree.Node  node = root;
    // left root right 
        while (true) {
            if(node != null){
                stack.push(node);
                node = node.left;
            }else{
                if(stack.isEmpty()){
                    break;
                }
                node = stack.pop();
                ans.add(node.data);
                node = node.right;
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        
        BinaryTree.Node root  = BinaryTree.createTree();
    }

}

File: LEETCODE/trees/IterativeTraversals/S3_InorderIteration.cpp
#include<bits/stdc++.h>
using namespace std;

struct Node{
    int val;
    Node* left;
    Node* right;

    Node(int val){
        this->val = val;
        this->left = nullptr;
        this->right = nullptr;

    }
};
vector<int> inorder(Node* root){
    stack<Node*> s;

    auto node = root;
    vector<int> ans;

    while (true){
        if(node != nullptr){
            s.push(node);
            node = node->left;
        }else{
            if(s.empty()){
                break;
            }
            node = s.top();
            s.pop();
            ans.push_back(node->val);
            node = node->right;
        }
    }
    
}
int main(){
    return 0;
}